File: agrotech-back/src/main/java/br/com/agrotech/agrotech_api/service/LeituraService.java
package br.com.agrotech.agrotech_api.service;

import br.com.agrotech.agrotech_api.model.Leitura;
import br.com.agrotech.agrotech_api.repository.LeituraRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import java.util.List;
import java.util.Optional;

/**
 * Classe de serviço para controle das regras de negócio da telemetria e leituras dos sensores.
 */
@Service
public class LeituraService {

    @Autowired
    private LeituraRepository leituraRepository;

    /**
     * Retorna o histórico de todas as capturas de sensores realizadas no sistema.
     *
     * @return {@link List} preenchida com o histórico completo de {@link Leitura}.
     */
    public List<Leitura> buscarTodas() {
        return leituraRepository.findAll();
    }

    /**
     * Busca um registro de telemetria individual através do seu ID de captura.
     *
     * @param id Identificador da medição.
     * @return Um {@link Optional} contendo a {@link Leitura} recuperada.
     */
    public Optional<Leitura> buscarPorId(Long id) {
        return leituraRepository.findById(id);
    }

    /**
     * Armazena uma nova leitura enviada pelos sensores de campo (pH, umidade, temperatura).
     *
     * @param leitura Os dados brutos de telemetria medidos pelo hardware IoT.
     * @return A {@link Leitura} salva e datada no banco de dados.
     */
    public Leitura salvar(Leitura leitura) {
        return leituraRepository.save(leitura);
    }

    /**
     * Permite a correção ou atualização manual de um registro de leitura dos sensores.
     *
     * @param id Identificador do registro de leitura a ser atualizado.
     * @param leituraAtualizada Objeto contendo os novos dados métricos de telemetria.
     * @return A {@link Leitura} modificada e persistida.
     * @throws RuntimeException Se a leitura informada não estiver gravada no banco.
     */
    public Leitura atualizar(Long id, Leitura leituraAtualizada) {
        return leituraRepository.findById(id)
                .map(leitura -> {
                    leitura.setPhSolo(leituraAtualizada.getPhSolo());
                    leitura.setUmidadeSolo(leituraAtualizada.getUmidadeSolo());
                    leitura.setTemperaturaAr(leituraAtualizada.getTemperaturaAr());
                    leitura.setDataHoraCaptura(leituraAtualizada.getDataHoraCaptura());
                    if (leituraAtualizada.getDispositivo() != null) {
                        leitura.setDispositivo(leituraAtualizada.getDispositivo());
                    }
                    return leituraRepository.save(leitura);
                })
                .orElseThrow(() -> new RuntimeException("Leitura não encontrada com o ID: " + id));
    }

    /**
     * Exclui um registro do histórico de medições de sensores.
     *
     * @param id Identificador do registro de leitura a ser excluído.
     * @throws RuntimeException Se o registro de leitura não for encontrado.
     */
    public void deletar(Long id) {
        if (!leituraRepository.existsById(id)) {
            throw new RuntimeException("Leitura não encontrada com o ID: " + id);
        }
        leituraRepository.deleteById(id);
    }
}

File: agrotech-front/src/services/leituraService.ts
import { api } from './api';
import type { Leitura } from '../types/Leitura';

/**
 * Serviço responsável pela comunicação de dados de telemetria e leituras dos sensores de campo.
 */
export const leituraService = {
    /**
     * Busca o histórico completo de capturas de dados realizadas pelos sensores.
     *
     * @returns {Promise<Leitura[]>} Uma promessa contendo o histórico de leituras.
     */
    listarTodas: async (): Promise<Leitura[]> => {
        const response = await api.get<Leitura[]>('/leituras');
        return response.data;
    },

    /**
     * Consulta os dados detalhados de uma leitura individual por ID.
     *
     * @param {number} id - Identificador único da medição.
     * @returns {Promise<Leitura>} Uma promessa contendo a leitura recuperada.
     */
    buscarPorId: async (id: number): Promise<Leitura> => {
        const response = await api.get<Leitura>(`/leituras/${id}`);
        return response.data;
    },

    /**
     * Envia para persistência um novo registro de telemetria medido no campo.
     *
     * @param {Leitura} leitura - Dados de temperatura, umidade e pH do solo.
     * @returns {Promise<Leitura>} Uma promessa contendo a leitura gravada no banco.
     */
    registrar: async (leitura: Leitura): Promise<Leitura> => {
        const response = await api.post<Leitura>('/leituras', leitura);
        return response.data;
    },

    /**
     * Permite a atualização manual de um registro de telemetria existente.
     *
     * @param {number} id - Identificador do registro a ser corrigido.
     * @param {Leitura} leitura - Dados atualizados de medição.
     * @returns {Promise<Leitura>} Uma promessa com a leitura corrigida.
     */
    atualizar: async (id: number, leitura: Leitura): Promise<Leitura> => {
        const response = await api.put<Leitura>(`/leituras/${id}`, leitura);
        return response.data;
    },

    /**
     * Remove um registro do histórico de telemetria do sistema.
     *
     * @param {number} id - Identificador da leitura a ser excluída.
     * @returns {Promise<void>} Promessa resolvida após a exclusão.
     */
    deletar: async (id: number): Promise<void> => {
        await api.delete(`/leituras/${id}`);
    },
};